lib/saluki-config/src/refresher.rs

```diff
@@ -0,0 +1,107 @@
+use std::sync::Arc;
+
+use crate::{ConfigurationError, GenericConfiguration};
+use arc_swap::ArcSwap;
+use reqwest;
+use saluki_error::GenericError;
+use serde::{de::DeserializeOwned, Deserialize};
+use serde_json::Value;
+use tokio::time::{sleep, Duration};
+
+const DATADOG_AGENT_CONFIG_ENDPOINT: &str = "https://localhost:5004/config/v1/";
+
+/// Configuration for setting up `RefreshableConfiguration`.
+#[derive(Default, Deserialize)]
+pub struct RefresherConfiguration {
+    auth_token_file_path: String,
+}
+
+/// The most recent configuration retrieved from the datadog-agent.
+#[derive(Default)]
+pub struct RefreshableConfiguration {
+    token: String,
+    values: ArcSwap<Value>,
+}
+
+impl RefresherConfiguration {
+    /// Creates a new `RefresherConfiguration` from the given configuration.
+    pub fn from_configuration(config: &GenericConfiguration) -> Result<Self, GenericError> {
+        Ok(config.as_typed()?)
+    }
+    /// Create `RefreshableConfiguration` from `RefresherConfiguration`
+    pub fn build(&self) -> Result<RefreshableConfiguration, GenericError> {
+        let raw_bearer_token = std::fs::read_to_string(&self.auth_token_file_path)?;
+        Ok(RefreshableConfiguration {
+            token: raw_bearer_token,
+            values: ArcSwap::from_pointee(serde_json::Value::Null),
+        })
+    }
+}
+impl RefreshableConfiguration {
+    /// Start a task that queries the datadog-agent config endpoint every 15 seconds.
+    pub fn spawn_refresh_task(self: Arc<Self>) {
+        tokio::spawn(async move {
+            loop {
+                match self.query_agent().await {
+                    Ok(_) => {
+                        // todo()!
+                    }
+                    Err(_e) => {
+                        // todo()!
+                    }
+                }
+                sleep(Duration::from_secs(15)).await; // Wait for 15 seconds
+            }
+        });
+    }
+    /// Query the datadog-agent config endpoint for the latest config
+    pub async fn query_agent(&self) -> Result<(), GenericError> {
```

Comment:
I would actually move this all into the refresh task, so that you're constructing the client outside of the loop, and then doing the request inside the loop.